### PR TITLE
Remove dependency on Doctrine/Cache

### PR DESCRIPTION
**Pull request by Kevinrob**

### CODE REVIEW

**Comment by Kevinrob** [comment]
@Thinkscape what do you think of this PR? 

- No dependency to `Doctrine/Cache`
- Better name of class (Middleware <-> Manager <-> Storage)
- `DoctrineCacheWrapper` for wrapping `Doctrine/Cache`

File: src/Storage/CacheStorageInterface.php [hunk 1]
```diff
@@ -0,0 +1,30 @@
+<?php
+/**
+ * Created by IntelliJ IDEA.
+ * User: Kevin
+ * Date: 12.08.2015
+ * Time: 19:07
+ */
+
```
Comment (Thinkscape) [inline src/Storage/CacheStorageInterface.php]: I'd remove that :)

**Comment by Thinkscape** [comment]
Sweet. btw: why do you call the outer manager "PrivateCache" ?

**Comment by Thinkscape** [comment]
Looking at your comment here https://github.com/Kevinrob/guzzle-cache-middleware/issues/27#issuecomment-130366617 private cache kinda becomes the "when-to-cache" strategy ... maybe then keep with:

1. CachingMiddleware - because guzzle 6 calls them middleware
2. CachingStrategy(Interface) - determines when to cache
3. StorageInterface - what holds the data, can be an adapter for some 3rd party caching framework or could be a real storage adapter.

The idea is to have at most 3 levels of nesting. The above allows for changing caching strategies and adapters. Whatcha think ?

**Comment by Thinkscape** [comment]
Some examples of strategies:

1. `CacheAlwaysStrategy implements CachingStrategy`
2. `RevalidateStrategy implements CachingStrategy`
3. `MimeMatchCacheStrategy implements CachingStrategy`


See also  https://github.com/Guzzle3/cache and https://github.com/Guzzle3/plugin-cache

File: src/Storage/CacheStorageInterface.php (hunk 1 above)
Comment (Kevinrob) [inline src/Storage/CacheStorageInterface.php]: Yeah, you're right!

**Comment by Kevinrob** [comment]
It's exactly when I would to create it.
`Private` is for a private cache (see RFC), maybe I will write the public after (shared cache).

Do you think `[...]Strategy` is better than `[...]Manager`?

**Comment by Thinkscape** [comment]
Yes. 
What RFC ? You mean #25  ?

Honestly, I haven't seen "private/public" cache distinction too often. It's confusing to me, as there's multiple ways to consider what is shared and how is it shared and on what level is it shared. "PrivateCache" is confusing to me on its own - cache by definition is not "private" within the app, unless you mean something like class-cache (i.e. `class foo { private $cache = []; }`). Otherwise it makes little sense.

If what you meant by "private" is how the backend is used, then you're probably not separating concerns here. The backend class would be of type (or configuration) that implies shared or not shared infrastructure. It has nothing to do with WHEN something is cached (for example APC cache is not shared, while memcached is usually shared across hosts).

A "Storage" is where the data is being stored.

A "Strategy" implies decision process, hence it makes more sense for stuff like pragmas, "cache always" or other ways to decide when to cache. The word "Manager" could be used, but I feel it doesn't convey the whole message ... for example, "CacheAlwaysManager" vs "CacheAlwaysStrategy" --- manager seems like a broader term, for example in some frameworks CacheManager could be a class to manage all caches (i.e. you have 10 backends and 20 instances, all managed by the Manager).

**Comment by Kevinrob** [comment]
You know, I will add some custom strategy like `CacheAlways` and other. But my priority is to make a functional [RFC 7234](https://tools.ietf.org/html/rfc7234) compliance strategy. And the RFC define two type of cache:
- **Public** => a shared cache that can cache HTTP resource considered as "public".
- **Private** => a private cache that can cache HTTP resource with private or specify to a user content too.

The custom policies are dangerous, because it is the server to define how the resource can be cached. For this, the servers must add cache directives to the response. All this is standard with the RFC.
It's the only way to be sure that you don't cache something that will create strange behavior.

The ultimate goal of this library is to be transparent. A developer can have performance without hard configuration and without downside.

**Comment by Thinkscape** [comment]
We had this kind of discussion already with Guzzle 3.  It's up to the developer to decide on the caching. They might want to go with RFC 7234, they might want to be more aggressive or not cache at all. Furthermore, there's tons of http hosts that are buggy or send bogus cache headers. 

Think of it as more of a library than a browser. RFC conformance makes sense more to something that's B2C in nature, like a web browser, which users rarely mess with. I do agree, that having a "DefaultCachingStrategy" (RFC 7234 compliant) is a useful thing, but Guzzle is not a browser - merely a smart curl wrapper, so I tend to say that this is more of a perk, or extra bonus, on top of "regular" strategies and revalidation settings that existed for years in Guzzle < 4.0 (and other php and non-php libraries).

**Comment by Kevinrob** [comment]
@Thinkscape 
Thank you for taking time to explain your point of view. It's interesting.
If the HTTP servers around the world are not sending good headers like you're saying, so I am sad about it!

I understand your thought about that. Now that the 3 layers (Middleware/Strategy/Storage) are correctly separated, it will be easy to add others strategies.
Will you use this library for yours projects?

**Comment by Thinkscape** [comment]
Yep, but I need full control over caching, same as I have with other libraries (not guzzle related).

Forget about "http servers sending headers". You're describing different use case. Guzzle is not a browser, it's a library. The developer must be in control, not remote host. Like mentioned before, we can have a "default strategy" that goes along some rfc rules, but other than that, there's a need for easy to use cache/nocache strategies.

**Pull request merged**